lib/source-code/source-code.js

```diff
@@ -972,6 +1051,91 @@ class SourceCode extends TokenStore {
 
     }
 
+    /**
+     * Traverse the source code and return the steps that were taken.
+     * @returns {Array<TraversalStep>} The steps that were taken while traversing the source code.
+     */
+    traverse() {
+
+        // Because the AST doesn't mutate, we can cache the steps
+        if (this.#steps) {
+            return this.#steps;
+        }
+
+        const steps = this.#steps = [];
+
+        /*
+         * This logic works for any AST, not just ESTree. Because ESLint has allowed
+         * custom parsers to return any AST, we need to ensure that the traversal
+         * logic works for any AST.
+         */
+        const emitter = createEmitter();
+        let analyzer = {
+            enterNode(node, parent) {
+                steps.push(new TraversalStep({
+                    type: "visit",
+                    target: node,
+                    phase: 1,
+                    args: [node, parent]
+                }));
+            },
+            leaveNode(node, parent) {
+                steps.push(new TraversalStep({
+                    type: "visit",
+                    target: node,
+                    phase: 2,
+                    args: [node, parent]
+                }));
+            },
+            emitter
+        };
+
+        /*
+         * We do code path analysis for ESTree only. Code path analysis is not
+         * necessary for other ASTs, and it's also not possible to do for other
+         * ASTs because the necessary information is not available.
+         *
+         * Generally speaking, we can tell that the AST is an ESTree if it has a
+         * Program node at the top level. This is not a perfect heuristic, but it
+         * is good enough for now.
+         */
+        const isESTree = this.ast.type === "Program";
+
+        if (isESTree) {
+            analyzer = new CodePathAnalyzer(analyzer);
+
+            CODE_PATH_EVENTS.forEach(eventName => {
+                emitter.on(eventName, (...args) => {
+                    steps.push(new TraversalStep({
+                        type: "call",
+                        target: eventName,
+                        args
+                    }));
+                });
+            });
+        }
+
+        /*
+         * The actual AST traversal is done by the `Traverser` class. This class
+         * is responsible for walking the AST and calling the appropriate methods
+         * on the `analyzer` object, which is appropriate for the given AST.
+         */
+        Traverser.traverse(this.ast, {
+            enter(node, parent) {
+
+                // save the parent node on a property for backwards compatibility
+                node.parent = parent;
+
+                analyzer.enterNode(node);
+            },
+            leave(node) {
+                analyzer.leaveNode(node);
+            },
+            visitorKeys: this.visitorKeys
+        });
```

Comment:
Are we sure that the current implementation of code path analysis doesn't expect `parent` property to be pre-set on all nodes (including not yet traversed ones)?

Reply:
I can only go by all of the tests that are passing. I'm assuming if there was a timing issue, that something would break.